### ISSUE TITLE
Compilation issues with CMake < 3.6 and 3.22

### ISSUE THREAD

**Issue by carlosperate** [lancaster-university/microbit-v2-samples#38]
Based on this technically the project should be 3.3+ compatible:
https://github.com/lancaster-university/microbit-v2-samples/blob/7f6be8079edf37a52d633add57ad01a3ccbced76/CMakeLists.txt#L22

However, on Windows any version below 3.6 fails to build, but it works in macOS and Ubuntu. In this example using 3.5.2:
- https://github.com/carlosperate/microbit-v2-samples/runs/4007178834?check_suite_focus=true
- https://github.com/carlosperate/microbit-v2-samples/runs/4007179052?check_suite_focus=true
```
[143/215] Building CXX object libraries/codal-microbit-v2/CMakeFiles/codal-microbit-v2.dir/source/MicroBitFlash.cpp.obj
FAILED: libraries/codal-microbit-v2/CMakeFiles/codal-microbit-v2.dir/source/MicroBitFlash.cpp.obj 
C:\Users\runneradmin\AppData\Local\Temp\tmp-3068SR2ZH1wX9d1e\gcc-7-2017-q4\bin\arm-none-eabi-g++.exe    -I../libraries -I../libraries/codal-microbit-v2/inc/bluetooth -I../libraries/codal-microbit-v2/inc/compat -I../libraries/codal-microbit-v2/inc -I../libraries/codal-microbit-v2/model -I../libraries/codal-nrf52/inc -I../libraries/codal-nrf52/inc/cmsis -I../libraries/codal-nrf52/nrfx/mdk -I../libraries/codal-nrf52/nrfx -I../libraries/codal-nrf52/nrfx/templates -I../libraries/codal-nrf52/nrfx/templates/nRF52833 -I../libraries/codal-nrf52/nrfx/drivers/include -I../libraries/codal-nrf52/nrfx/drivers/src -I../libraries/codal-core/./inc/core -I../libraries/codal-core/./inc/driver-models -I../libraries/codal-core/./inc/drivers -I../libraries/codal-core/./inc/streams -I../libraries/codal-core/./inc/types -I../libraries/codal-microbit-nrf5sdk/../codal-nrf52/nrfx/hal -I../libraries/codal-microbit-nrf5sdk/sdk_config -I../libraries/codal-microbit-nrf5sdk/nRF5SDK_mods -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/pwr_mgmt -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/sortlist -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/strerror -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/softdevice/common -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/crc32 -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/bootloader/dfu -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/util -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/ble/common -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/balloc -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/ble/peer_manager -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/ringbuf -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/timer -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/log -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/ble/nrf_ble_gatt -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/bootloader -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/fstorage -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/experimental_section_vars -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/softdevice/s113/headers -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/mutex -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/delay -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/bootloader/ble_dfu -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/atomic_fifo -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/atomic -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/memobj -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/fds -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/atomic_flags -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/softdevice/s113/headers/nrf52 -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/ble/ble_services/ble_dfu -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/external/fprintf -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/svc -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/libraries/log/src -I../libraries/codal-microbit-nrf5sdk/nRF5SDK/components/ble/ble_services/ble_dis -Wno-expansion-to-defined -mcpu=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=softfp -fno-exceptions -fno-unwind-tables -ffunction-sections -fdata-sections -Wall -Wextra -Wno-unused-parameter -std=c++11 -fwrapv -fno-rtti -fno-threadsafe-statics -fno-exceptions -fno-unwind-tables -Wl,--gc-sections -Wl,--sort-common -Wl,--sort-section=alignment -Wno-array-bounds -include D:/a/microbit-v2-samples/microbit-v2-samples/build/codal_extra_definitions.h -ID:/a/microbit-v2-samples/microbit-v2-samples/utils/cmake/toolchains/ARM_GCC -O2 -g -DNDEBUG   -DAPP_TIMER_V2 -DAPP_TIMER_V2_RTC1_ENABLED -DNRF_DFU_TRANSPORT_BLE=1 -DSOFTDEVICE_PRESENT -DNRF52833_XXAA -DNRF52833 -DTARGET_MCU_NRF52833 -DNRF5 -DNRF52833 -D__CORTEX_M4 -DS113 -DTOOLCHAIN_GCC -D__START=target_start -MMD -MT libraries/codal-microbit-v2/CMakeFiles/codal-microbit-v2.dir/source/MicroBitFlash.cpp.obj -MF libraries/codal-microbit-v2/CMakeFiles/codal-microbit-v2.dir/source/MicroBitFlash.cpp.obj.d -o libraries/codal-microbit-v2/CMakeFiles/codal-microbit-v2.dir/source/MicroBitFlash.cpp.obj -c ../libraries/codal-microbit-v2/source/MicroBitFlash.cpp
../libraries/codal-microbit-v2/source/MicroBitFlash.cpp: In member function 'void MicroBitFlash::erase_page(uint32_t*)':
../libraries/codal-microbit-v2/source/MicroBitFlash.cpp:119:9: error: 'ble_running' was not declared in this scope
     if (ble_running())
         ^~~~~~~~~~~
../libraries/codal-microbit-v2/source/MicroBitFlash.cpp: In member function 'void MicroBitFlash::flash_burn(uint32_t*, uint32_t*, int)':
../libraries/codal-microbit-v2/source/MicroBitFlash.cpp:161:9: error: 'ble_running' was not declared in this scope
     if (ble_running())
         ^~~~~~~~~~~
[144/215] Building CXX object libraries/codal-microbit-v2/CMakeFiles/codal-microbit-v2.dir/source/bluetooth/MicroBitLEDService.cpp.obj
[145/215] Building CXX object libraries/codal-microbit-v2/CMakeFiles/codal-microbit-v2.dir/source/MicroBitFileSystem.cpp.obj
ninja: build stopped: subcommand failed.
```

Also, the latest release, v3.22.0-rc1, fails on Windows as well:
- https://github.com/carlosperate/microbit-v2-samples/runs/4007178956?check_suite_focus=true
- https://github.com/carlosperate/microbit-v2-samples/runs/4007179122?check_suite_focus=true
```
[213/215] Linking CXX executable MICROBIT
FAILED: MICROBIT libMICROBIT.dll.a 
cmd.exe /C "cd . && C:\Users\runneradmin\AppData\Local\Temp\tmp-3452F6ug5gi7I0HB\gcc-10-2020-q4\gcc-arm-none-eabi-10-2020-q4-major\bin\arm-none-eabi-g++.exe -std=c++11 -fwrapv -fno-rtti -fno-threadsafe-statics -fno-exceptions -fno-unwind-tables -Wl,--gc-sections -Wl,--sort-common -Wl,--sort-section=alignment -Wno-array-bounds -Wl,--no-wchar-size-warning -Wl,--gc-sections -Wl,--wrap,atexit -mcpu=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=softfp -Wl,--start-group -lstdc++ -lsupc++ -lm -lc -lgcc -lnosys -Wl,--end-group -T"D:/a/microbit-v2-samples/microbit-v2-samples/libraries/codal-microbit-v2/ld/nrf52833.ld" -Wl,-Map,MICROBIT.map -Wl,--start-group CMakeFiles/MICROBIT.dir/source/main.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/AccelerometerTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/AudioTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/BLETest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/BlinkyTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/ButtonTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/CapTouchTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/DeepSleepTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/DisplayTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/GPIOTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/LightLevelTestRaw.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/MicrophoneTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/NoiseProfiler.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/OOB.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/PowerManagementTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/RadioTestRx.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/SerialNumber.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/SerialStreamer.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/SpeakerTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/TemperatureTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/USBFlashTest.cpp.obj  libcodal-microbit-v2.a  libcodal-core.a  libcodal-nrf52.a  libcodal-microbit-nrf5sdk.a  libcodal-nrf52.a  libcodal-core.a  ../libraries/codal-microbit-v2/lib/bootloader.o  ../libraries/codal-microbit-v2/lib/mbr.o  ../libraries/codal-microbit-v2/lib/settings.o  ../libraries/codal-microbit-v2/lib/softdevice.o  ../libraries/codal-microbit-v2/lib/uicr.o  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -lnosys -lstdc++ -lsupc++ -lm -lc -lgcc -lstdc++ -lsupc++ -lm -lc -lgcc -Wl,--end-group  --specs=nano.specs -o MICROBIT && cd ."
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lkernel32
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -luser32
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lgdi32
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lwinspool
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lshell32
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lole32
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -loleaut32
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -luuid
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lcomdlg32
c:/users/runneradmin/appdata/local/temp/tmp-3452f6ug5gi7i0hb/gcc-10-2020-q4/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -ladvapi32
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by carlosperate** [comment]
Latest version 3.22.1 has the same problem:
https://github.com/carlosperate/microbit-v2-samples/runs/4820719389?check_suite_focus=true
```
[213/215] Linking CXX executable MICROBIT
FAILED: MICROBIT libMICROBIT.dll.a 
cmd.exe /C "cd . && C:\Users\runneradmin\gcc-arm-none-eabi-0.0.0-win32\gcc-arm-none-eabi-10.3-2021.10\bin\arm-none-eabi-g++.exe -std=c++11 -fwrapv -fno-rtti -fno-threadsafe-statics -fno-exceptions -fno-unwind-tables -Wl,--gc-sections -Wl,--sort-common -Wl,--sort-section=alignment -Wno-array-bounds -Wl,--no-wchar-size-warning -Wl,--gc-sections -Wl,--wrap,atexit -mcpu=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=softfp -Wl,--start-group -lstdc++ -lsupc++ -lm -lc -lgcc -lnosys -Wl,--end-group -T"D:/a/microbit-v2-samples/microbit-v2-samples/libraries/codal-microbit-v2/ld/nrf52833.ld" -Wl,-Map,MICROBIT.map -Wl,--start-group CMakeFiles/MICROBIT.dir/source/main.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/AccelerometerTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/AudioTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/BLETest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/BlinkyTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/ButtonTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/CapTouchTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/DeepSleepTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/DisplayTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/GPIOTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/LightLevelTestRaw.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/MicrophoneTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/NoiseProfiler.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/OOB.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/PowerManagementTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/RadioTestRx.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/SerialNumber.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/SerialStreamer.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/SpeakerTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/TemperatureTest.cpp.obj CMakeFiles/MICROBIT.dir/source/samples/USBFlashTest.cpp.obj  libcodal-microbit-v2.a  libcodal-core.a  libcodal-nrf52.a  libcodal-microbit-nrf5sdk.a  libcodal-nrf52.a  libcodal-core.a  ../libraries/codal-microbit-v2/lib/bootloader.o  ../libraries/codal-microbit-v2/lib/mbr.o  ../libraries/codal-microbit-v2/lib/settings.o  ../libraries/codal-microbit-v2/lib/softdevice.o  ../libraries/codal-microbit-v2/lib/uicr.o  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -lnosys -lstdc++ -lsupc++ -lm -lc -lgcc -lstdc++ -lsupc++ -lm -lc -lgcc -Wl,--end-group  --specs=nano.specs -o MICROBIT && cd ."
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lkernel32
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -luser32
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lgdi32
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lwinspool
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lshell32
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lole32
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -loleaut32
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -luuid
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -lcomdlg32
c:/users/runneradmin/gcc-arm-none-eabi-0.0.0-win32/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.exe: cannot find -ladvapi32
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Error: Process completed with exit code 1.
```

**Comment by carlosperate** [comment]
I just remembered that it's possible Windows is the only platform using ninja, so the problem might lay there? Just a random guess.

The ninja version used in the latest CI run was 1.10.2.3, which is the latest release at the time of writting (v1.10.2).

**Comment by microbit-carlos** [comment]
@JohnVidler 3.3 was released in 2015 and 3.6.0 was release in 2016, more than 6 years ago https://github.com/Kitware/CMake/commit/e31084e65745f9dd422c6aff0a2ed4ada6918805.

There is not much point trying to resolve the Windows issue with 3.3 to 3.5, and because of that issue we haven't even been testing those versions in the other OS in CI either, so let's move the min version to 3.6, which we are currently testing in CI.

Our main constrain is making sure the project is compatible with v3.8, as that's what's currently inside the MakeCode pxt yotta docker images.

**Comment by JohnVidler** [comment]
Changed in upstream codal as well, to align with 3.6 here.
